### ISSUE TITLE
Allow variable merge with namespace when value of variable is a allowed merging declaration

### ISSUE THREAD

**Issue by FilipeBeck** [microsoft/TypeScript#45975]
# Suggestion

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

It is not allowed merging namespace with variables. But, in some cases, it should be when the value itself is a allowed merging.

This is allowed:
```typescript
function foo() {
  // ...
}
namespace foo {
  export const bar = 'bar'
}

class Foo {
  // ...
}
namespace Foo {
  export const bar = 'bar'
}
```
But this is not:
```typescript
const foo = function() {
  // ...
}
namespace foo {
  export const bar = 'bar'
}

const Foo = class {
  // ...
}
namespace Foo {
  export const bar = 'bar'
}
```
As `class A {...` and `const A = class {...` are equivalent, it should compile with no erros.

**Comment by andrewbranch** [comment]
Is there a scenario where you need to use `const A = class` instead of `class A`?

**Comment by FilipeBeck** [comment]
With `class`, no until now, but I put it to give it more scope. It's common for me to have problems with _function type_ vars, like here: https://github.com/FilipeBeck/react-audio-engine/blob/develop/src/reconciler/Components.ts#L83-L96

I have:
```typescript
const Scene: React.FC< ...
namespace Scene { ...
```

And I cannot change to:
```typescript
function Scene(props: { children ...
```

Because `Scene` is the return of `React.forwardRef`

**Comment by andrewbranch** [comment]
But I don’t think that helps you, because we couldn’t know that the return of `React.forwardRef` _is_ safely mergeable with anything else. The only possible way I can see this being allowed (and I’m still unsure about it) is if the initializer is a function expression or class expression, which means you _would_ be able to rewrite it as a function declaration or class declaration.

**Comment by FilipeBeck** [comment]
Yes, I'm realizing it now. It is not possible to guarantee that the _return value_ does not have other properties than those declared in the interface, as long as it respects the interface.

> The only possible way I can see this being allowed (and I’m still unsure about it) is if the initializer is a function expression ...

Maybe, it can still be useful for scenarios with arrow functions (although it doesn't solve the problem that made me open the suggestion)

**Comment by FilipeBeck** [comment]
In [this code](https://github.com/FilipeBeck/react-audio-engine/blob/develop/src/reconciler/Components.ts#L83-L96), I was trying to re-export a `const enum` and, as `const enum` does not generate code, I solved it this way:
```typescript
export const Scene: React.FC<Scene.Props> & { LatencyCategory: typeof ATOM.Scene.LatencyCategory } = (
  React.forwardRef((props, ref) => {
    return React.createElement(Tag.SCENE, { ...props, ref }, props.children)
  }) as any
)
```

This works even if the variable has a property with the same name, as seen [here](https://www.typescriptlang.org/play?#code/MYewdgzgLgBApmArgWxgGQIZQcAngYSzgHMQAnXGAbwCgYYAhAQTSYDl8BRAERgF4YAcgBGGADYYwwOABNBAGjowAkmwAqnAEpN8a5QDVO-IQEsw2MhmBQTANzgKlABVYBNZvgDSxwQAcJuKLAANaOAL40NGYWAGZWcDAAyhjI-nCaRNRKltgAXDBIyMJwZEqY2FIERKQU+dBkZsQ0ETQxiFI24DBkcFCIZGCJIMhwyali6UQAFACU+WNpGdhZ9D19Ayv03UT5ACy7AIwADEeKW+hElYTYNbj5gvvHR4JKES2gkLBDIwsTS3DzFKLTIAMmoFwqeGuJHIdxgUFwvjgIBiEJwVRusJgYWMa36g2GoyBf2mMxgGAg5LAuEiHwgIAmADoxCBiFNBN8ieNJthGTkHPIkoTfjy4HyiDMaHSGWKWWy-GQQEiyAihT9iaLGeV0dDbgo1VzgbztVdqrDJdKmXL2ZyRf8tZcoWaKPrbRr7SanZiKIzmKwODwZkA) 

Would allowing merge with `const enum` violate the design goals?

**Comment by FilipeBeck** [comment]
Maybe I might not have expressed myself well before because my english is _more google than me_. Sorry by the possible language mistakes.

**Comment by FilipeBeck** [comment]
If it's useful for someone else, better discuss it in a separate and more specific issue.
